### PR TITLE
minor: Use `CargoConfig.extra_args` for fetching metadata

### LINKED ISSUE #14510: rust-analyzer.cargo.extraArgs not working
**rust-analyzer version**: (eg. output of "rust-analyzer: Show RA Version" command, accessible in VSCode via <kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>)
0.3.1463-standalone and 0.4.1465-standalone

**rustc version**: (eg. output of `rustc -V`)
rustc 1.70.0-nightly (2eaeb1eee 2023-04-05)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTC`, `RUSTUP_HOME` or `CARGO_HOME`)

I would like to pass some extra arguments to cargo.
According to the rust-analyser documentation this should be possible via `rust-analyzer.cargo.extraArgs` which says _Extra arguments that are passed to every cargo invocation._

**settings.json:**
```
{
    "[rust]": {
        "editor.defaultFormatter": "rust-lang.rust-analyzer",
        "editor.formatOnSave": true
        
    },
    "rust-analyzer.cargo.extraArgs": ["-Z", "bindeps"],
}
```

The flag is used to get this working:
```
[dependencies]
wasi_snapshot_preview1 = { git = "https://github.com/bytecodealliance/preview2-prototyping", rev = "083879c", target = "wasm32-unknown-unknown", artifact = "cdylib" }
```

Unfortunately rust-analyser shows me an error. From this error message it also appears that `-Z bindeps` is not passed with.

```
[ERROR rust_analyzer::main_loop] FetchWorkspaceError:
rust-analyzer failed to load workspace: Failed to read Cargo metadata from Cargo.toml file /home/giom/repos/jfc/noops/Cargo.toml, Some(Version { major: 1, minor: 70, patch: 0, pre: Prerelease("nightly") }): Failed to run `cd "/home/giom/repos/jfc/noops" && "cargo" "metadata" "--format-version" "1" "--manifest-path" "/home/giom/repos/jfc/noops/Cargo.toml" "--filter-platform" "x86_64-unknown-linux-gnu"`: `cargo metadata` exited with an error: error: failed to load manifest for workspace member `/home/giom/repos/jfc/noops/noops-server`

Caused by:
  failed to parse manifest at `/home/giom/repos/jfc/noops/noops-server/Cargo.toml`

Caused by:
  `artifact = …` requires `-Z bindeps` (wasi_snapshot_preview1)
```
If I use the command from the error message and append `-Z bindeps` the metadata is created successfully.

### PR DESCRIPTION
**Pull request by jakhh8**
Fixes #14510

### CODE REVIEW

File: crates/project-model/src/cargo_workspace.rs [hunk 1]
```diff
@@ -297,6 +297,7 @@ impl CargoWorkspace {
             let other_options: Vec<_> = targets
                 .into_iter()
                 .flat_map(|target| ["--filter-platform".to_string(), target])
+                .chain(config.extra_args.clone())
```
Comment (Veykril) [inline crates/project-model/src/cargo_workspace.rs:300]: This is only set when the targets are empty here, we should put this after the `!targets.is_empty()` branch

Comment (jakhh8) [reply crates/project-model/src/cargo_workspace.rs]: Should be fixed now

**Comment by utkarshgupta137** [comment]
This change broke a workaround many of us have been using to have a separate directory for rust-analyzer: https://github.com/rust-lang/rust-analyzer/issues/6007#issuecomment-1480600503.
I don't think this change is a good idea because it limits the arguments in `cargo.extraArgs` to only those which are compatible with `cargo-metadata`. `cargo-metadata` supports a small subsection of the arguments supported by other commands such as `cargo-check`. In particular, I think the following arguments could be useful, which are no longer allowed: `package`, `exclude`, `jobs`, `profile`, `target-dir`, `release`.
If we want to keep this, then at least the incompatible arguments should be removed from `cargo-metadata`, or a brute-force fallback without extraArgs could also work. I would also note that the OP of the issue was able to workaround his problem using a `.cargo/config` file.

**Comment by Veykril** [comment]
https://github.com/rust-lang/rust-analyzer/pull/14712 should patch over this. Unstable flags seem to be the only relevant ones so thats all we pass  through then

**Pull request merged**